exercises/variables/variables5.rs

```diff
@@ -6,4 +6,9 @@ fn main() {
     println!("Spell a Number : {}", number);
     let number: u32 = 3; // don't rename this variable
     println!("Number plus two is : {}", number + 2);
+    {
+        let number: f32 = 0.1;
+        println!("Number plus 1 is : {}", number + 1.0)
```

Comment:
This shadowed variable mechanism is extremely dangerous.

Mutability also doesn't seem a big concern when providing type.

I am not certain I understand the concept of immutability here in theory. I would assume that the languages symbol table would not allow for a let to allow for a rebinding. I can only assume that the nature of let deals with malloc and helps to release old bindings regardless of the symbol in question. I know rust has great memory management so I wonder if it is doing ARC at runtime or just dealloc and malloc at the let step. On exit it is doing normal GC.

Reply:
Tell me why you think it's a concern. I'm kinda ambivalent about it; the whole "multi-purpose-single-variable" thing I seen in Erlang/Elixir to (the use arity and scope [AFAICT] to delineate uniqueness) so there is precedence to some degree, but I'm not totally sold on it either.

Reply:
I think its dependent on how you use these closures. I love the idea of creating a nested scope but I also consider thats the domain of a function or a lambda.

The risk I see is in the reading of the code, in a sufficiently complex file I might not notice that an object has been changed in place for a local action.

It reminds me a little of JS hoisting https://www.digitalocean.com/community/tutorials/understanding-hoisting-in-javascript

albeit that is an unexpected compilation behavior and this is an intentional act of the user. Gotta see some more use cases to understand it better.

I get it a variable is ultimately an address where values live so reusing it isn't such a concern to the computer executing but might be some thinking for a user reading.

Its all very interesting and I do enjoy having something new to think about.